proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+type NestedEtcdStatus struct {
+    // Ready is set if all resources have been created
+    Ready bool `json:"ready,omitempty"`
+
+    // ProviderObject is the object that used by the underlying provider.
+    // Under the native mode, the ProviderObject points to the StatefulSet
+    // that contains the component's pods.
+    ProviderObject *corev1.ObjectReference `json:referralObject,omitempty`
+
+    // EtcdDomain defines how to address the etcd instance
+    Addresses []NestedEtcdAddresses `json:"addresses,omitempty"`
+}
+
+// NestedEtcdAddresses defines the observed addresses for etcd
+type NestedEtcdAddresses struct {
+    // IP Address of the etcd instance.
+    // +optional
+    IP string `json:"ip,omitempty"`
+    
+    // Hostname of the etcd instance
+    Hostname string `json:"hostname,omitempty"`
+    
+    // Port of the etcd instance
+    // +optional
+    Port int32 `json:"port"`
+}
+
+// NestedEtcd is the Schema for the nestedetcds API
+type NestedEtcd struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    
+    Spec   NestedEtcdSpec   `json:"spec,omitempty"`
+    Status NestedEtcdStatus `json:"status,omitempty"`
+}
+```
+
+### NestedKAS CRD
+
+```go
+type NestedKASSpec struct {
+    // The NestedControlPlane that owns this NestedKAS
+    NestedControlPlane *corev1.ObjectReference `json:nestedControlPlane,omitempty`
+   
+    // NestedComponentSpec contains the common and user-specified information that are 
+    // required for creating the component
+    // +optional
+    NestedComponentSpec NestedComponentSpec  `json:nestedComponentSpec,omitempty`
+
+    // Provider defines which provider will be used to create the kube-apiserver 
+    // +optional
+    Provider KASProvider `json:provider,omitempty`
+}
+
+// NestedKASStatus defines the observed state of NestedKAS
+type NestedKASStatus struct {
+    // Ready is set if all resources have been created
+    // +kubebuilder:default=false
+    Ready bool `json:"ready,omitempty"`
+
+    // ProviderObject is the object that used by the underlying provider.
+    // Under the native mode, the ProviderObject points to the StatefulSet
+    // that contains the component's pods.
+    ProviderObject *corev1.ObjectReference `json:referralObject,omitempty`
+
+    // EtcdDomain defines how to address the etcd instance
+    // +optional
+    KASService *corev1.Service `json:"addresses,omitempty"`
+}
+
+// NestedKAS is the Schema for the NestedKASs API
+type NestedKAS struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    
+    Spec   NestedKASSpec   `json:"spec,omitempty"`
+    Status NestedKASStatus `json:"status,omitempty"`
+}
+```
+
+### NestedKCM CRD
+
+```go
+// NestedKCMSpec defines the desired state of NestedKCM
+type NestedKCMSpec struct {
+    // The NestedControlPlane that owns this NestedKCM
+    NestedControlPlane *corev1.ObjectReference `json:nestedControlPlane,omitempty`
```

Comment:
```suggestion
    ControlPlaneRef *corev1.ObjectReference `json:nestedControlPlane,omitempty`
```

Reply:
Will remove the ControlPlaneRef, as we are going to use the OwnerReference